pkg/kf/commands/routes/create_route.go

```diff
@@ -0,0 +1,149 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routes
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"path"
+
+	"github.com/GoogleCloudPlatform/kf/pkg/kf/commands/config"
+	"github.com/GoogleCloudPlatform/kf/pkg/kf/internal/routeutil"
+	"github.com/GoogleCloudPlatform/kf/pkg/kf/routes"
+	"github.com/knative/pkg/apis/istio/common/v1alpha1"
+	networking "github.com/knative/pkg/apis/istio/v1alpha3"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NewCreateRouteCommand creates a CreateRoute command.
+func NewCreateRouteCommand(
+	p *config.KfParams,
+	c routes.Client,
+) *cobra.Command {
+	var hostname, urlPath string
+
+	cmd := &cobra.Command{
+		Use:   "create-route DOMAIN [--hostname HOSTNAME] [--path PATH]",
+		Short: "Create a route",
+		Example: `
+  # Using namespace (instead of SPACE)
+  kf create-route example.com --hostname myapp # myapp.example.com
+  kf create-route -n myspace example.com --hostname myapp # myapp.example.com
+  kf create-route example.com --hostname myapp --path /mypath # myapp.example.com/mypath
+
+  # [DEPRECATED] Using SPACE to match 'cf'
+  kf create-route myspace example.com --hostname myapp # myapp.example.com
+  kf create-route myspace example.com --hostname myapp --path /mypath # myapp.example.com/mypath
+  `,
+		Args: cobra.RangeArgs(1, 2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			space, domain := p.Namespace, args[0]
+			if len(args) == 2 {
+				space = args[0]
+				domain = args[1]
+				fmt.Fprintln(cmd.OutOrStderr(), `
+[WARN]: passing the SPACE as an argument is deprecated.
+Instead use the --namespace flag.`)
+			}
+
+			if p.Namespace != "" && p.Namespace != space {
+				return errors.New("SPACE (argument) and namespace (if provided) must match")
+			}
+
+			if hostname == "" {
+				return errors.New("--hostname is required")
+			}
+
+			cmd.SilenceUsage = true
+
+			var pathMatchers []networking.HTTPMatchRequest
+			if urlPath != "" {
+				pathMatchers = append(pathMatchers, networking.HTTPMatchRequest{
+					URI: &v1alpha1.StringMatch{
+						Prefix: path.Join("/", urlPath),
+					},
+				})
+			}
+
+			hostDomain := hostname + "." + domain
+
+			vs := &networking.VirtualService{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "networking.istio.io/v1alpha3",
+					Kind:       "VirtualService",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: routeutil.EncodeRouteName(hostname, domain, urlPath),
```

Comment:
since these are kind of getting lost here, I think it's important to add the host, domain and path as annotations to the VirtualService so it's easier to grok when you're digging through by hand.